### PR TITLE
Better toDate/toDateTime functions.

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Provide monotonicity for `toDate/toDateTime` functions in more cases. Now the input arguments are saturated more naturally and provides better monotonicity.

Detailed description / Documentation draft:

This PR is considered as a breaking change since the SQL result such as `toDate(-1)`, `toDate(1000000000)` is different. But it's just more natural and should be self-evident.

### CODE REVIEW

File: src/Functions/FunctionsConversion.h [hunk 1]
```diff
@@ -1412,6 +1506,25 @@ struct ToNumberMonotonicity
...
+        auto which = WhichDataType(type);
+        if (which.isDateOrDateTime() || which.isInt8() || which.isInt16() || which.isUInt8() || which.isUInt16())
+            return {true, true, true};
+        else if (
+            (which.isUInt() && ((left.isNull() || left.get<UInt64>() < 0xFFFF) && (right.isNull() || right.get<UInt64>() >= 0xFFFF)))
+            || (which.isInt() && ((left.isNull() || left.get<Int64>() < 0xFFFF) && (right.isNull() || right.get<Int64>() >= 0xFFFF)))
+            || (which.isFloat() && ((left.isNull() || left.get<Float64>() < 0xFFFF) && (right.isNull() || right.get<Float64>() >= 0xFFFF))))
+            return {};
+        else
+            return {true, true, true};
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsConversion.h:1524]: What happens for String, Decimal...?

**Comment by alexey-milovidov** [comment]
Need to check/update tests...

File: src/Functions/FunctionsConversion.h (hunk 1 above)
Comment (amosbird) [reply src/Functions/FunctionsConversion.h:1525]: Decimal and other complex types cannot be converted to date/datetime. And it's unlikely a valid use case...

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@amosbird Let's check, maybe the task https://github.com/ClickHouse/ClickHouse/issues/10977 is also resolved by this PR?

**Pull request merged**